src/bounce_solution/bounce_solution.cpp

```diff
@@ -0,0 +1,1256 @@
+    if (SolutionList.size() == 1)
+    {
+      // Failed to find the solution
+      SetBounceSol();
+      return;
+    };
+  }
+
+  GWScanTowardsLowAction();
+  GWScanTowardsHighAction();
+
+  std::size_t NumOfSol = SolutionList.size();
+
+  for (int i = 0; i < 2; i++)
+  {
+    std::vector<double> nextTList;
+
+    for (std::size_t sol = 0; sol < SolutionList.size() - 1; sol++)
+    {
+      if (SolutionList[sol].action / SolutionList[sol].T > 200) continue;
+      if (SolutionList[sol + 1].action / SolutionList[sol + 1].T < 50) continue;
+
+      double NumOfT =
+          ceil((SolutionList[sol + 1].action / SolutionList[sol + 1].T -
+                SolutionList[sol].action / SolutionList[sol].T) /
+               20);
+
+      double dT = (SolutionList[sol + 1].T - SolutionList[sol].T) / NumOfT;
+      for (int it_T = 1; it_T < NumOfT; it_T++)
+      {
+        nextTList.push_back(SolutionList[sol].T + it_T * dT);
+      }
+    }
+
+    for (auto j : nextTList)
+    {
+      CalculateActionAt(j);
+    }
+
+    if (NumOfSol == SolutionList.size()) break;
+    NumOfSol = SolutionList.size();
+  }
+  GWScanTowardsLowAction();
+  GWScanTowardsHighAction();
+  SetBounceSol();
+}
+
+void BounceSolution::GWScanTowardsHighAction()
+{
+  for (int i = 0;
+       i <=
+       1. + (200 - SolutionList.back().action / SolutionList.back().T) / 5.;
+       i++)
+  {
+    if (SolutionList.back().action / SolutionList.back().T > 200) break;
+
+    // Try to calculate until s3/T = 200
+    double t1            = SolutionList[SolutionList.size() - 2].T;
+    double t2            = SolutionList[SolutionList.size() - 1].T;
+    double s1            = SolutionList[SolutionList.size() - 2].action / t1;
+    double s2            = SolutionList[SolutionList.size() - 1].action / t2;
+    double goal          = s2 + 5;
+    std::size_t NumOfSol = SolutionList.size();
+
+    double T = ((s1 - goal) * t2 - (s2 - goal) * t1) / (s1 - s2);
+
+    // Action is not monotonic
+    if (T < t2) return;
+
+    if (T > this->Tc) // Action is flat
+    {
+      if (t2 + (t2 - t1) > this->Tc) return; // Close already
+      CalculateActionAt(t2 + (t2 - t1));
+      i--;
+    }
+    else
+    {
+      CalculateActionAt(T);
+      if (NumOfSol == SolutionList.size())
+        CalculateActionAt(
+            T, false); // No solution was found. Calculate using dumb method
+    }
+    if (NumOfSol == SolutionList.size()) return; // No solution was found. Abort
+  }
+}
+
+void BounceSolution::GWScanTowardsLowAction()
+{
+  for (int i = 0;
+       i <=
+       1. + (SolutionList.front().action / SolutionList.front().T - 50) / 10.;
+       i++)
+  {
+    if (SolutionList.front().action / SolutionList.front().T < 50) break;
+    // Try to calculate it at s3/T = 300
+    double t1            = SolutionList[0].T;
+    double t2            = SolutionList[1].T;
+    double s1            = SolutionList[0].action / t1;
+    double s2            = SolutionList[1].action / t2;
```

Comment:
General note: To access elements in a vector it is better to use `.at(i)` instead of `[i]` for the out of bounds check

Reply:
These cases are already protected. From my understanding ```.at(i)``` is slower than ```[i]``` so if we are sure the index is within bounds we can use ```[i]```, particularly for heavy computations. 
If you are okay with it, I will still use "[i]" when we are sure we are within bounds.

Reply:
Yes than that is fine